### PR TITLE
[Merged by Bors] - feat: port GroupTheory.Subsemigroup.Centralizer

### PR DESCRIPTION
**Pull request by hargoniX**
I commented out the assertion in the bottom since we don't have it as of now and it is not relevant to the port itself.

### CODE REVIEW

File: Mathlib/GroupTheory/Subsemigroup/Centralizer.lean [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+/-!
+# Centralizers of magmas and semigroups
+
+## Main definitions
+
+* `set.centralizer`: the centralizer of a subset of a magma
+* `subsemigroup.centralizer`: the centralizer of a subset of a semigroup
+* `set.add_centralizer`: the centralizer of a subset of an additive magma
+* `add_subsemigroup.centralizer`: the centralizer of a subset of an additive semigroup
```
Comment (dupuisf) [inline Mathlib/GroupTheory/Subsemigroup/Centralizer.lean:22]:
```suggestion
* `Set.centralizer`: the centralizer of a subset of a magma
* `Subsemigroup.centralizer`: the centralizer of a subset of a semigroup
* `Set.addCentralizer`: the centralizer of a subset of an additive magma
* `AddSubsemigroup.centralizer`: the centralizer of a subset of an additive semigroup
```

File: Mathlib/GroupTheory/Subsemigroup/Centralizer.lean [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+We provide `monoid.centralizer`, `add_monoid.centralizer`, `subgroup.centralizer`, and
+`add_subgroup.centralizer` in other files.
```
Comment (dupuisf) [inline Mathlib/GroupTheory/Subsemigroup/Centralizer.lean:25]:
```suggestion
We provide `Monoid.centralizer`, `AddMonoid.centralizer`, `Subgroup.centralizer`, and
`AddSubgroup.centralizer` in other files.
```

File: Mathlib/GroupTheory/Subsemigroup/Centralizer.lean [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+
+variable {M : Type _} {S T : Set M}
+
+namespace Set
+
+variable (S)
+
+/-- The centralizer of a subset of a magma. -/
+@[to_additive add_centralizer " The centralizer of a subset of an additive magma. "]
```
Comment (dupuisf) [inline Mathlib/GroupTheory/Subsemigroup/Centralizer.lean:36]:
```suggestion
@[to_additive addCentralizer " The centralizer of a subset of an additive magma. "]
```

**Review by dupuisf — COMMENTED** [review]
Just some naming adjustments and `#align` statements to add, otherwise LGTM.

bors d+

File: Mathlib/GroupTheory/Subsemigroup/Centralizer.lean [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+def centralizer [Mul M] : Set M :=
+  { c | ∀ m ∈ S, m * c = c * m }
+#align set.centralizer Set.centralizer
```
Comment (dupuisf) [inline Mathlib/GroupTheory/Subsemigroup/Centralizer.lean:39]:
```suggestion
  { c | ∀ m ∈ S, m * c = c * m }
#align set.centralizer Set.centralizer
#align set.add_centralizer Set.addCentralizer
```
`to_additive` doesn't add `#align` statements automatically, so these have to be added manually throughout the file as well.

Comment (hargoniX) [reply Mathlib/GroupTheory/Subsemigroup/Centralizer.lean:39]: How am I supposed to do the alignment for things such as:
```
@[to_additive "The centralizer of a subset of an additive semigroup."]
def centralizer : Subsemigroup M where
  carrier := S.centralizer
  mul_mem' := Set.mul_mem_centralizer
#align subsemigroup.centralizer Subsemigroup.centralizer
```
There is no explicit naming done here, what is the course of action in this case?

Comment (hargoniX) [reply Mathlib/GroupTheory/Subsemigroup/Centralizer.lean:39]: @dupuisf ?

Comment (ADedecker) [reply Mathlib/GroupTheory/Subsemigroup/Centralizer.lean:39]: Guessing the generated name, checking that it is indeed generated, and using that in the `#align` should work.

**Pull request closed without merging**